### PR TITLE
Support for binary decompression of dynamic patches.

### PR DESCRIPTION
**Pull request by sbaranov**
Also see https://github.com/flutter/flutter/pull/27754 for compression support in the flutter tool.

### CODE REVIEW

File: shell/platform/android/io/flutter/view/ResourceExtractor.java [hunk 1]
```diff
@@ -339,4 +346,114 @@ private String checkTimestamp(File dataDir) {
...
+    private static void copy(InputStream in, OutputStream out) throws IOException {
+        byte[] buf = new byte[16 * 1024];
+        for (int i; (i = in.read(buf)) >= 0; ) {
+            out.write(buf, 0, i);
+        }
+    }
+
+    private static void read(InputStream in, byte[] buf, int off, int len) throws IOException {
+        for (int i, n = 0; n < len; n += i) {
+            if ((i = in.read(buf, off+n, len-n)) < 0) {
```
Comment (jason-simmons) [inline shell/platform/android/io/flutter/view/ResourceExtractor.java]: nit: add spaced around the `+` and `-` operators

File: shell/platform/android/io/flutter/view/ResourceExtractor.java [hunk 2]
```diff
@@ -339,4 +346,114 @@ private String checkTimestamp(File dataDir) {
...
+                throw new IOException("Unexpected EOF");
+            }
+        }
+    }
+
+    // This is a Java port of algorithm from Flutter framework bsdiff.dart.
```
Comment (jason-simmons) [inline shell/platform/android/io/flutter/view/ResourceExtractor.java]: Move this to a separate file

File: shell/platform/android/io/flutter/view/ResourceExtractor.java (hunk 1 above)
Comment (sbaranov) [reply shell/platform/android/io/flutter/view/ResourceExtractor.java]: Done

File: shell/platform/android/io/flutter/view/ResourceExtractor.java (hunk 2 above)
Comment (sbaranov) [reply shell/platform/android/io/flutter/view/ResourceExtractor.java]: Done

**Comment by sbaranov** [comment]
PTAL

**Comment by dnfield** [comment]
This commit broke the devicelab.  https://logs.chromium.org/logs/flutter/buildbucket/cr-buildbucket.appspot.com/8921783973932371216/+/steps/build_android_profile_gen_snapshot/0/stdout

**Pull request merged**